### PR TITLE
feat(schemas): declare the five stats.nba.com raw payload families

### PR DESCRIPTION
**Pull request by saiemgilani**
Declares the raw payload shapes for `stats.nba.com` / `stats.wnba.com`, extending the seven ESPN families already in `sportsdataverse/schemas/`.

## The envelope is not uniform — there are five families

The working assumption for this task was a single `{resultSets: [{name, headers, rowSet}]}` shape. Probing the committed `hoopR-nba-stats-raw` tree (**489,977 files, 64 endpoint dirs**) found five, and the assumed one covers a minority of the archive:

| schema | shape | endpoints |
|---|---|---:|
| `nba_stats_result_sets` | classic; `resultSets` is a **list** | 35 |
| `nba_stats_result_set` | **singular** key, dict value | 3 |
| `nba_stats_result_sets_grouped` | plural key but **dict** value, and `headers` is a list of column-**group** dicts (`columnNames`/`columnSpan`/`columnsToSkip`) rather than name strings | 2 |
| `nba_stats_v3` | modern `{<entity>, meta}` tree — **no `resultSets` at all** | 13 |
| `nba_stats_v3_period` | period-number map of v3 payloads | 1 |

`nba_stats_v3` is the **bulk of the tree**: `boxscoretraditionalv3` and `playbyplayv3` are 39,244 files each, and the v3 box scores plus pbp and the schedule are ~340k of the 490k payloads. A single result-sets schema would have described roughly 20% of the archive and silently ignored the rest.

Two edges worth calling out:

- **`resultSet` vs `resultSets`** is a genuinely different key, not a list of length one. Reading a singular payload with the plural path finds nothing. A test pins that they are not interchangeable.
- **The grouped family's `headers` are dicts.** A parser assuming flat string headers mis-reads the shot-zone column bands entirely. Related latent hazard documented in the schema: `season_capture._ids_from()` iterates `payload.get("resultSets") or []` and calls `.get()` on each element — against this family that iterates the dict's *keys* (strings) and would `AttributeError`. It doesn't fire today only because `_ids_from` is pointed exclusively at family-A payloads.

## New DSL knob: `required_any`

The v3 entity key is named per endpoint (`boxScoreTraditional`, `game`, `leagueSchedule`, `scoreboard`, …). Requiring only `meta` would let a payload whose body was dropped validate clean; a schema per entity key would be a dozen near-identical files. `required_any` says "at least one of these keys must be present" — six lines in `validate_payload`.

## Validated against the real archive, not the inline samples

8,608 payloads sampled across **54 endpoint directories** were auto-classified from their keys and validated:

```
checked            : 10319
empty ({})         : 1711
classified         :  8608
unclassified       :     0
VALIDATION FAILURES:     0
```

Zero unclassified means the five families are exhaustive for this archive; zero failures means each one describes what it claims to.

## The `{}` assertion is load-bearing

The suite pins that an empty payload fails **every** family, because the same probe found a real corruption:

> **3,347 committed files are exactly `{}`.** Ten endpoints are **100% empty** — `playergamelogs` (553), `teamgamelogs` (550), `playercompare` (555), `matchupsrollup` (80), `leaguedashptteamdefend` (88) and five `draftcombine*` (31 each) — plus **71% of both shot-locations endpoints** (1,239 files).

They persist because that scraper's `write_payload()` has no guard and its resume check is `path.exists()` — **presence, not content** — so one empty write is never retried and a backfill no-ops.

`{}` is not a shape the API produces: an endpoint with no rows still returns the full envelope with `rowSet: []` (see `leagueleaders/1996/playoffs_pergame.json`). So these are failed fetches that were persisted as if they were data.

Note the discriminator is `size <= 2`, not "small": `playbyplayv3/1996/0029500001.json` is a valid 201-byte envelope whose `game.actions` is `[]` — v3 pbp genuinely has no actions that far back, and that file is correct.

Repair of the archive is tracked separately in the `-raw` repo; this PR gives it the detector.

## Verification

```
uv run pytest tests/test_raw_schemas.py  -> 57 passed
uv run ruff check                        -> All checks passed
uv run python tools/codegen/generate.py --check -> all generated files current
```

## Summary by Sourcery

Add raw schema support for the five distinct stats.nba.com / stats.wnba.com payload families and enforce their validation semantics, including detection of corrupted empty captures.

New Features:
- Introduce schemas for nba_stats_result_sets, nba_stats_result_set, nba_stats_result_sets_grouped, nba_stats_v3, and nba_stats_v3_period raw payloads.
- Add support for schema-level required_any key groups to validate v3 envelopes with endpoint-specific entity keys.

Enhancements:
- Document multi-envelope behavior for stats.nba.com providers and map each new schema to its corresponding raw file pattern.
- Extend validate_payload to enforce required_any groups while preserving tolerance for unknown keys.

Tests:
- Add comprehensive tests covering all nba_stats families, including acceptance of real-shaped samples, rejection of empty payloads, key non-interchangeability, v3 entity-key requirements, and period completeness/OT handling.

## Summary by CodeRabbit

* **New Features**
  * Added validation support for five NBA statistics payload formats, including classic, singular, grouped, modern v3, and period-based responses.
  * Added support for required regulation periods and optional overtime periods.
  * Improved detection of incomplete, malformed, or incorrectly structured nested payloads while allowing additional fields.

* **Documentation**
  * Documented supported NBA payload structures and validation requirements.

* **Tests**
  * Added comprehensive NBA and WNBA coverage for valid, invalid, empty, singular, plural, grouped, v3, and overtime payloads.

### CODE REVIEW

File: sportsdataverse/schemas/__init__.py [hunk 1]
```diff
@@ -181,6 +199,15 @@ def validate_payload(name: str, payload: Any) -> list[str]:
...
         problems.extend(_check(name, key, payload[key], declared))
+    # `required_any`: at least one of these keys must be present. The
+    # stats.nba.com v3 endpoints all ship {<entityKey>, meta} where the entity
+    # key is named per endpoint (boxScoreTraditional, game, leagueSchedule,
+    # scoreboard, ...). Requiring only `meta` would let a payload whose body
+    # was dropped validate clean, and one schema per entity key would be a
+    # dozen near-identical files.
+    for group in schema.get("required_any") or []:
+        if not any(key in payload for key in group):
+            problems.append(f"{name}: payload has none of {sorted(group)!r}; expected one")
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply sportsdataverse/schemas/__init__.py]: Fixed in e553e4d7. A bare string is now treated as a one-key group, so `required_any: [foo]` behaves as the author obviously meant rather than testing the payload for single characters. Test added (`test_required_any_tolerates_a_bare_string_group`).

File: sportsdataverse/schemas/raw/nba_stats_result_sets_grouped.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+# sampled payloads. Observed on leaguedashplayershotlocations and
+# leaguedashteamshotlocations, e.g. leaguedashteamshotlocations/1996/
+# playoffs_base_pergame.json -- headers is a 2-element list of group dicts over
+# a 26-wide rowSet.
+#
+# Known latent hazard: season_capture._ids_from() iterates
+# `payload.get("resultSets") or []` and calls .get() on each element. Against
+# this family that iterates the dict's KEYS (strings) and would raise
+# AttributeError. It does not fire today only because _ids_from is pointed
+# exclusively at nba_stats_result_sets payloads.
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply sportsdataverse/schemas/raw/nba_stats_result_sets_grouped.yaml:18]: Agreed, and fixed properly rather than left as a note — in [hoopR-nba-stats-raw](https://github.com/sportsdataverse/hoopR-nba-stats-raw/pull/4), where `_ids_from` actually lives (commit `5193544`).

It now goes through a `_result_tables()` normaliser handling all three table-bearing shapes: `resultSets` as a list, `resultSets` as a dict, and the **singular** `resultSet` — which that helper was previously blind to entirely, not just fragile against. v3 payloads carry no result tables and yield `[]`. Grouped headers are skipped safely (they're column-group dicts, so they cannot match a column name) and the row index is bounded so a short row can't `IndexError`. Five tests cover the shapes.

Your framing was the right one: modelling the envelopes is exactly what makes it likelier someone points a generic helper at the wrong one.

File: tests/test_raw_schemas.py [hunk 3]
```diff
@@ -162,3 +162,138 @@ def test_required_keys_match_what_was_observed(name, required):
...
+def test_v3_requires_an_entity_body_not_just_meta():
+    """A dropped body must not validate clean. `meta` alone is a truncated
+    payload, and every v3 endpoint pairs it with exactly one entity key."""
+    meta_only = {"meta": {"request": "http://...", "time": "t", "version": 1}}
+    problems = validate_payload("nba_stats_v3", meta_only)
+    assert problems and "expected one" in problems[0]
+
+
+@pytest.mark.parametrize("entity", ["game", "leagueSchedule", "scoreboard", "boxScoreUsage"])
+def test_v3_accepts_each_observed_entity_key(entity):
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply tests/test_raw_schemas.py]: Done in e553e4d7 — parametrized from `load_raw_schema("nba_stats_v3")["required_any"][0]`, so a newly-declared entity key is covered the moment it lands in the YAML.

File: sportsdataverse/schemas/raw/nba_stats_v3_period.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+# Periods 1-4 are required: every completed NBA game has four. Overtimes are
+# optional and open-ended -- 5 through 14 covers ten OTs, well past the
+# six-OT record. A game missing period 3 is a partial capture, which is
+# precisely what this schema exists to catch.
+additional_properties: true
+required:
+  "1": dict
+  "2": dict
+  "3": dict
+  "4": dict
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply sportsdataverse/schemas/raw/nba_stats_v3_period.yaml:26]: Fixed in e553e4d7 — this was the most substantive comment on the PR and you're right that it defeated the schema's purpose: `{"1": {}, "2": {}, "3": {}, "4": {}}` is exactly a capture that lost every payload body, which is the corruption these schemas exist to catch.

Added a `values_of` key to the DSL. `nba_stats_v3_period` declares `values_of: nba_stats_v3`, and each period is validated as a full v3 envelope with the offending period named in the message (`nba_stats_v3_period[3] -> nba_stats_v3: payload has none of [...]`).

Two scoping decisions worth flagging:
- It applies only to **declared** keys (periods 1–14), so `additional_properties: true` keeps meaning the same thing everywhere. Validating unknown keys too would report a new top-level field as a malformed period — caught by an existing test.
- A schema naming itself as `values_of` raises instead of recursing.

Three tests: the all-empty map, a single truncated period, and a non-dict period.
